src/DssProxyActionsCharter.sol

```diff
@@ -0,0 +1,693 @@
+        if (nib > 0) {
+            (, uint256 rate,,,) = VatLike(vat).ilks(ilk);
+            uint256 dtab = mul(rate, uint256(dart)); // rad
+            uint256 coin = mul(dtab, nib) / WAD;     // rad
+            wadN = sub(dtab, coin) / RAY;
+        } else {
+            wadN = wad;
+        }
+    }
+
+    function _getDrawDart(
+        address vat,
+        address jug,
+        bytes32 ilk,
+        uint256 wad
+    ) internal returns (int256 dart) {
+        // Updates stability fee rate
+        uint256 rate = JugLike(jug).drip(ilk);
+
+        // Gets DAI balance of the urn in the vat
+        uint256 dai = VatLike(vat).dai(address(this));
+
+        // If there was already enough DAI in the vat balance, just exits it without adding more debt
+        if (dai < mul(wad, RAY)) {
+            // Calculates the needed dart so together with the existing dai in the vat is enough to exit wad amount of DAI tokens
+            dart = toInt256(sub(mul(wad, RAY), dai) / rate);
+            // This is needed due lack of precision. It might need to sum an extra dart wei (for the given DAI wad amount)
+            dart = mul(uint256(dart), rate) < mul(wad, RAY) ? dart + 1 : dart;
+        }
+    }
+
+    function _getWipeDart(
+        address vat,
+        uint256 dai,
+        address urn,
+        bytes32 ilk
+    ) internal view returns (int256 dart) {
+        // Gets actual rate from the vat
+        (, uint256 rate,,,) = VatLike(vat).ilks(ilk);
+        // Gets actual art value of the urn
+        (, uint256 art) = VatLike(vat).urns(ilk, urn);
+
+        // Uses the whole dai balance in the vat to reduce the debt
+        dart = toInt256(dai / rate);
+        // Checks the calculated dart is not higher than urn.art (total debt), otherwise uses its value
+        dart = uint256(dart) <= art ? - dart : - toInt256(art);
+    }
+
+    function _getWipeAllWad(
+        address vat,
+        address urn,
+        bytes32 ilk
+    ) internal view returns (uint256 wad) {
+        // Gets actual rate from the vat
+        (, uint256 rate,,,) = VatLike(vat).ilks(ilk);
+        // Gets actual art value of the urn
+        (, uint256 art) = VatLike(vat).urns(ilk, urn);
+        // Gets actual dai amount in the urn
+        uint256 dai = VatLike(vat).dai(address(this));
+
+        uint256 rad = sub(mul(art, rate), dai);
+        wad = rad / RAY;
+
+        // If the rad precision has some dust, it will need to request for 1 extra wad wei
+        wad = mul(wad, RAY) < rad ? wad + 1 : wad;
+    }
+
+    // Public functions
+
+    function transfer(address gem, address dst, uint256 amt) public {
+        GemLike(gem).transfer(dst, amt);
+    }
+
+    function ethJoin_join(address charter, address apt) public payable {
+        // Wraps ETH in WETH
+        GemJoinLike(apt).gem().deposit{value: msg.value}();
+        // Approves adapter to take the WETH amount
+        GemJoinLike(apt).gem().approve(charter, msg.value);
```

Comment:
We might want to cache the `gem` value instead of getting it twice. This is something that I'm going to fix for the proxy actions as well.

Reply:
Done in your suggested commit.